tests/multi_ptr/multi_ptr_implicit_conversions.h

```diff
@@ -0,0 +1,237 @@
+          //  that does not accept that type, but accepts some
+          //  other type T2; in particular:
+          //  <...>
+          //  - when initializing a new object of type T2, including return
+          //  statement in a function returning T2;
+          //  <...>
+          dest_multi_ptr_t mptr_dest = invoke_conversion_t{}(mptr_from);
+
+          // for cases, when dest_multi_ptr_t equals to multi_ptr<void>
+          T value_dest = static_cast<T>(*(mptr_dest.get()));
+
+          res_acc[0] = (value_dest == expected_val);
+        });
+      });
+    }
+    CHECK(res);
+  }
+
+ public:
+  void operator()(const std::string &type_name,
+                  const std::string &address_space_str,
+                  const std::string &is_decorated_str) {
+    SECTION(section_name("Verifying implicit conversion from "
+                         "multi_ptr<T,address_space,decorated>")
+                .with("T", type_name)
+                .with("address_space", address_space_str)
+                .with("decorated", is_decorated_str)
+                .create()) {
+      SECTION("Implicit conversion from multi_ptr<T> to multi_ptr<void>") {
+        using from_multi_ptr_t = sycl::multi_ptr<T, address_space, decorated>;
+
+        using dest_multi_ptr_not_decorated_t =
+            sycl::multi_ptr<void, address_space, sycl::access::decorated::no>;
+        using dest_multi_ptr_decorated_t =
+            sycl::multi_ptr<void, address_space, sycl::access::decorated::yes>;
+
+        SECTION("Conversion to multi_ptr<void, decorated::yes>") {
+          preform_implicit_conversion_test<from_multi_ptr_t,
+                                           dest_multi_ptr_decorated_t>();
+        }
+        SECTION("Conversion to multi_ptr<void, decorated::no>") {
+          preform_implicit_conversion_test<from_multi_ptr_t,
+                                           dest_multi_ptr_not_decorated_t>();
+        }
+      }
+      // clang-format off
+      SECTION(
+        "Implicit conversion from multi_ptr<const T> to multi_ptr<const void>") {
+        // clang-format on
+        using from_multi_ptr_t =
+            sycl::multi_ptr<const T, address_space, decorated>;
+
+        using dest_multi_ptr_not_decorated_t =
+            sycl::multi_ptr<const void, address_space,
+                            sycl::access::decorated::no>;
+        using dest_multi_ptr_decorated_t =
+            sycl::multi_ptr<const void, address_space,
+                            sycl::access::decorated::yes>;
+
+        SECTION("Conversion to multi_ptr<const void, decorated::yes>") {
+          preform_implicit_conversion_test<from_multi_ptr_t,
+                                           dest_multi_ptr_decorated_t>();
+        }
+        SECTION("Conversion to multi_ptr<const void, decorated::no>") {
+          preform_implicit_conversion_test<from_multi_ptr_t,
+                                           dest_multi_ptr_not_decorated_t>();
+        }
```

Comment:
We are iterating over different fields of `decorated` enum's, so can we use something like this?
```suggestion
        using dest_multi_ptr_t =
            sycl::multi_ptr<const void, address_space, decorated>;

            SECTION(section_name("Conversion to ")
                  .with("T", type_name)
                  .with("address_space", address_space_name)
                  .with("decorated", is_decorated_name)
                  .create()) {
       preform_implicit_conversion_test<from_multi_ptr_t, dest_multi_ptr_t>();
      }
```

Reply:
[Spec](https://www.khronos.org/registry/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:multiptr) says, that we `multi_ptr<T>` can be implicitly converted to `multi_ptr<void>` with decorated and non-decorated address space, despite source parameters of `multi_ptr`. By these checks I wanted to cover these two cases:
- `multi_ptr<T,space,decorated>` -> `multi_ptr<void,space,decorated::no>`
- `multi_ptr<T,space,decorated>` -> `multi_ptr<void,space,decorated::yes>`

That's why we need two different sections here, with `decorated::yes` and `decorated::no`